Enclave model upload optimization

itstauq opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

itstauq committed:
Optimization 1

itstauq committed:
Remove asset_list from model before saving it to the model stash

itstauq commented on packages/syft/src/syft/client/domain_client.py:
```diff
@@ -182,6 +182,7 @@ def upload_model(self, model: CreateModel) -> SyftSuccess | SyftError:
             return valid
 
         # Step 4. Upload Model to Model Stash
+        model.asset_list = []
```
@rasswanth-s the changes in `on_disk.py` alone reduces the `upload_model` time from ~1m20s to ~30s. Adding this line further reduces the total time to ~5s.

From what I gather (and I might be wrong here), we are already uploading these assets to the action store in step 2. Can we have some sort of lazy loading for this asset list instead, where they contain only the action ids, and the actual objects can be fetched on demand instead?

rasswanth-s replied:
This is because of the data and mock field in the asset, we could delete them after uploading to blob storage.
I will push a commit fixing the same.

rasswanth-s replied:
Done [2358ede](https://github.com/OpenMined/PySyft/pull/9053/commits/2358ede3116ac276d17b1ff62bcacd72b20160bc)

itstauq commented on packages/syft/src/syft/store/blob_storage/on_disk.py:
Earlier, most of the time was being spent on serde and signing of the very large assets. Directly writing the file onto the disk and bypassing the API call layer reduces the time for uploading GPT2 model (~500 MB) from 1m20s to ~30s. Great for prototyping. Potential long term fix would be using pre-signed URLs instead of API calls.

madhavajay replied:
We might want this is an env variable or something for local mode only, because otherwise this would trigger when you run a python web server that you want to run as a normal server right?

rasswanth-s replied:
I did a performance and found that , signing is the root bottleneck.

when using other libraries like `cryptography` it took under a second.

pynacl's, internal hashing method maybe slow.

rasswanth-s replied:
This fix is okay for now, 

This would fails for single container scenario, as the path is not local to a client.
but we could fix it later.
I think  real fix would be to optimize the signing part, for which I will create an issue shortly.
and also there are some performance overhead in pydantic v2.

rasswanth-s committed:
upload digital signature benchmarks

rasswanth-s committed:
upload digital signature benchmarks for ecc, rsa

rasswanth-s committed:
Add benchmarks also for pycryptodome

rasswanth-s committed:
clear up and data and mock references in asset before uploading to server , as they are referenced by id's

rasswanth-s committed:
clear up action data cache when uploading to blob store

rasswanth-s committed:
add pyinstrument profiling

rasswanth-s commented on packages/syft/src/syft/node/routes.py:
```diff
@@ -120,7 +120,7 @@ async def stream_upload(peer_uid: str, url_path: str, request: Request) -> Respo
         response_class=JSONResponse,
     )
-    def root() -> dict[str, str]:
+    async def root() -> dict[str, str]:
```
Pyinstrument requires async routes , as it currently does not work with sync routes.

rasswanth-s committed:
cleanup profile settings

rasswanth-s committed:
change default profile interval to be 0.001

rasswanth-s committed:
revert to sync routes

rasswanth-s committed:
shift pyinstrument to make_routes, as async routes are not fully supported,
disable pyinstrument middlware temporarily
re-name appsettings to serversettings

rasswanth-s reviewed (approved):
Great work @itstauq
